### PR TITLE
[Semi Auto] Entropy SPMD Rule

### PR DESCRIPTION
**Pull request by JZ-LIANG**
### PR types
Function optimization

### PR changes
Others

### Description

Pcard-70448

SPMD rule for entropy ops.
Support shard on broadcast axes and normalized axis.

### CODE REVIEW

File: paddle/fluid/distributed/auto_parallel/spmd_rules/cross_entropy_with_softmax_spmd_rule.cc [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+CrossEntropyWithSoftmaxSPMDRule::InferForward(
+    const std::vector<DistTensorSpec>& input_specs,
+    const paddle::framework::AttributeMap& attrs) {
+  // step0: verify input args based on cross_entropy_with_softmax logic
+  auto input_specs_size = input_specs.size();
+  PADDLE_ENFORCE_EQ(
+      input_specs_size,
+      2,
+      phi::errors::InvalidArgument("The size of InputSpec of cross entropy "
+                                   "with softmax should be 1, but got [%d].",
```
Comment (zhiqiu) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/cross_entropy_with_softmax_spmd_rule.cc]:
```suggestion
                                   "with softmax should be 2, but got [%d].",
```

File: test/auto_parallel/spmd_rules/test_cross_entropy_with_softmax_rule.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+from paddle.distributed.auto_parallel.static.completion import get_spmd_rule
+from paddle.distributed.auto_parallel.static.dist_attribute import (
+    DistTensorSpec,
+    TensorDistAttr,
+)
+from paddle.distributed.fleet import auto
+
+
+class TestEmbeddingSPMDRule(unittest.TestCase):
```
Comment (zhiqiu) [inline test/auto_parallel/spmd_rules/test_cross_entropy_with_softmax_rule.py]: CorssEntropy

File: paddle/fluid/distributed/auto_parallel/spmd_rules/rules.h [hunk 3]
```diff
@@ -145,6 +146,10 @@ REGISTER_SPMD_RULE(lookup_table_v2, EmbeddingSPMDRule);
 REGISTER_SPMD_RULE(softmax, SoftmaxSPMDRule);
 REGISTER_SPMD_RULE(log_softmax, SoftmaxSPMDRule);
 
+// cross_entropy_with_softmax
+REGISTER_SPMD_RULE(cross_entropy_with_softmax, CrossEntropyWithSoftmaxSPMDRule);
+REGISTER_SPMD_RULE(softmax_with_cross_entropy, CrossEntropyWithSoftmaxSPMDRule);
```
Comment (zhiqiu) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/rules.h:151]: The two ops have same attrs?

File: paddle/fluid/distributed/auto_parallel/spmd_rules/cross_entropy_with_softmax_spmd_rule.cc (hunk 1 above)
Comment (JZ-LIANG) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/cross_entropy_with_softmax_spmd_rule.cc]: done

File: test/auto_parallel/spmd_rules/test_cross_entropy_with_softmax_rule.py (hunk 2 above)
Comment (JZ-LIANG) [reply test/auto_parallel/spmd_rules/test_cross_entropy_with_softmax_rule.py]: done

File: paddle/fluid/distributed/auto_parallel/spmd_rules/rules.h (hunk 3 above)
Comment (JZ-LIANG) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/rules.h:151]: yes, here is to adopt for the difference between paddle static mode and dygraph mode, we will unify it in future.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**